### PR TITLE
[NFC] Batchnorm first solver

### PR DESCRIPTION
**Pull request by DrizztDoUrden**
The first forward solver for batch norm.
No performance issues have been found on my little 40 minutes bench.

### CODE REVIEW

**Comment by DrizztDoUrden** [comment]
Oh, wow, how did I miss the naming from the first commit...

**Comment by atamazov** [comment]
Please merge fresh develop and switch to clang-format-10

**Comment by DrizztDoUrden** [comment]
sure, I'm currently testing, would push merged soon, whether it would pass the current iteration or not.

**Comment by junliume** [comment]
@DrizztDoUrden The following tests FAILED:

[2021-08-05T17:17:37.052Z] 	  3 - test_bn_3d_spatial_test (Failed)

**Comment by DrizztDoUrden** [comment]
yeah, I am currently debugging

**Comment by junliume** [comment]
@DrizztDoUrden 9/81 Test  #2: **test_bn_3d_peract_test** .....................***Failed   44.24 sec

> [2021-08-06T14:27:11.000Z] Memory access fault by GPU node-2 (Agent handle: 0xc89f40) on address 0x7fe91c120000. Reason: Unknown.
> 
> [2021-08-06T14:27:11.000Z] CMake Error at test_test_bn_3d_peract_test.cmake:7 (message):
> 
> [2021-08-06T14:27:11.000Z]   Test failed

**Comment by DrizztDoUrden** [comment]
I was not able to reproduce the segfault but it didn't reproduce on Jenkins either.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -120,6 +120,8 @@ set( MIOpen_Source
     solver/activ/fwd_1.cpp
     solver/activ/bwd_0.cpp
     solver/activ/bwd_1.cpp
+    batchnorm/problem_description.cpp
+    solver/batchnorm/forward_per_activation_single.cpp
```
Comment (muralinr) [inline src/CMakeLists.txt:124]: Only forward batchnorm code reorg and other changes are done.  How about backward batchnorm?

File: src/batchnorm/problem_description.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
+/*******************************************************************************
```
Comment (muralinr) [inline src/batchnorm/problem_description.cpp:1]: We need to run batchnorm tests for correctness and performance with resnet and other network configs.  Are those tests done?  Also, we need to see any gaps in testing as changes are many..

File: src/ocl/batchnormocl.cpp [hunk 3]
```diff
@@ -37,6 +37,10 @@
 #include <miopen/convolution.hpp>
 #include <miopen/mlo_internal.hpp>
 #include <miopen/stringutils.hpp>
+#include <miopen/batchnorm/invoke_params.hpp>
```
Comment (muralinr) [inline src/ocl/batchnormocl.cpp:40]: Are these changes tested for both FP32 and FP16?

**Review by muralinr — COMMENTED** [review]
Overall changes look good.  We need to ensure sufficient FP16 and FP32 testing with various network configs.

File: src/CMakeLists.txt (hunk 1 above)
Comment (DrizztDoUrden) [reply src/CMakeLists.txt:124]: I make one solver per PR. Otherwise, this would be even more immense. This one is focused on the forward single kernel batchnorm.

File: src/ocl/batchnormocl.cpp (hunk 3 above)
Comment (DrizztDoUrden) [reply src/ocl/batchnormocl.cpp:40]: Yes.

File: src/batchnorm/problem_description.cpp (hunk 2 above)
Comment (DrizztDoUrden) [reply src/batchnorm/problem_description.cpp:1]: I have used a simple self-made test set + corner cases for different condition paths. My script multiplies parameter value sets, limits them by image size, and dumps a lot of data from logs like times, solver name and compilation details. That data is later automatically compared with similar data gathered from the develop. I would be happy to execute whatever to test if you would like.

<details>
  <summary>Batchnorm parameter values used in my core bench</summary>

  ```powershell
$DataTypes = @("", "fp16")
$Directions = @(0, 1, 2)
$Modes = @(0, 1)
$InputSizes = @(@(64, 128), @(64, 128), @(256, 256), @(720, 480), @(1280, 760), @(1920, 1080), @(3840, 2160))
$Channels = @(1, 16, 32, 50, 64, 128)
$BatchSizes = @(1, 10, 50, 100)
$backwardsPropation = @($false, $true)
$run = @($false, $true)
$save = @($false, $true)
$Iterations = 10
$Limit = 32Mb
# This limits data_type_size*W*H*batchsize*in_channels to make bench time sane
# 32Mb was used here because as with larger images multi-kernel paths gets invoked a lot, and corner cases for that were hand-picked later.
  ```

</details>

If you would like more details on the test script I could post full files somewhere.

Comment (muralinr) [reply src/batchnorm/problem_description.cpp:1]: Here are some configs for accuracy and performance tests. Can you please try with various batch sizes(n =32, 64,...1024)?  

FP32:
#!/bin/bash

./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 2048 -H 7 -W 7 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 1024 -H 14 -W 14 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 512 -H 7 -W 7 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 256 -H 14 -W 14 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 256 -H 56 -W 56 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 64 -H 112 -W 112 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 64 -H 56 -W 56 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 512 -H 28 -W 28 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnorm -m 1 -s 1 -r 1 -n ${1} -c 128 -H 28 -W 28 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0

FP16:
#!/bin/bash

./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 2048 -H 7 -W 7 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 1024 -H 14 -W 14 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 512 -H 7 -W 7 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 256 -H 14 -W 14 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 64 -H 112 -W 112 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 256 -H 56 -W 56 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 64 -H 56 -W 56 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 512 -H 28 -W 28 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0
./bin/MIOpenDriver bnormfp16 -m 1 -s 1 -r 1 -n ${1} -c 128 -H 28 -W 28 -V ${3} -F 1 -b 0 -i ${2} -t 1 -w 0

**Review by junliume — APPROVED** [review]

File: src/batchnorm/problem_description.cpp (hunk 2 above)
Comment (DrizztDoUrden) [reply src/batchnorm/problem_description.cpp:1]: In progress

Comment (DrizztDoUrden) [reply src/batchnorm/problem_description.cpp:1]: No significant difference

Comment (DrizztDoUrden) [reply src/batchnorm/problem_description.cpp:1]: I can post the full comparison if you want.

Comment (muralinr) [reply src/batchnorm/problem_description.cpp:1]: Good.

**Review by muralinr — APPROVED** [review]
As there is no significant performance difference, I approve this PR.

File: src/convolution.cpp [hunk 4]
```diff
@@ -510,9 +510,7 @@ std::size_t ConvolutionDescriptor::BackwardWeightsGetWorkSpaceSizeGEMM(
         return *std::max_element(gemm_ws_szs.begin(), gemm_ws_szs.end());
     }
 #else
-    std::ignore = dyDesc;
```
Comment (junliume) [inline src/convolution.cpp:513]: [Post Merge Review]
This issue was introduced in #852 and fixed with #1058
JIRA reported at https://ontrack-internal.amd.com/browse/SWDEV-301586 
Mainline: 
from 791b419c49485353c0e70ccb92be26bcdcf97e9c (05/18/2021)
fixed in 1e90891903ad795a3928cde443ed1921f9c9de52 (latest of as 09/02/2021)
